Specify protocol V3 when creating PackageSource in GetSymbolPackageUpdateResource

Zaneo opened the pull request:
<!-- DO NOT MODIFY OR DELETE THIS TEMPLATE. IT IS USED IN AUTOMATION. -->
## Bug

<!-- Search https://github.com/NuGet/Home/issues, and create one if you can't find a suitable issue. -->
<!-- Paste the full link, like https://github.com/nuget/home/issues/1000. GitHub will render is neatly. -->
Fixes: https://github.com/NuGet/Home/issues/12951

Regression? Last working version:

## Description
In `GetSymbolPackageUpdateResource` if the requested `source` cannot be found in the current `sourceProvider` a new `PackageSource` is created. Then using V3 requests the `SymbolPackageUpdateResourceV3` is attempted to be retreived. This will obviously fail on a V2 package source.

This changes the creation of `PackageSource` to use `ProtocolV3`.

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [ ] Automated tests added
  - **OR**
  <!-- Describe why you haven't added automation. -->
  - [ ] Test exception
  - **OR**
  - [x] N/A <!-- Infrastructure, documentation etc. -->

- **Documentation**
  <!-- Please link the PR/issue if appropriate -->
  - [ ] Documentation PR or issue filled
  - **OR**
  - [x] N/A

Zaneo committed:
Specify protocol when creating PackageSource

Since the PackageSource will be queried for V3 resources ( SymbolPackageUpdateResourceV3 ) it should be created as a V3 PackageSource.

Zaneo commented:
@dotnet-policy-service agree

zivkan commented on src/NuGet.Core/NuGet.Commands/Utility/CommandRunnerUtility.cs:
```diff
@@ -106,7 +106,11 @@ public static async Task<SymbolPackageUpdateResourceV3> GetSymbolPackageUpdateRe
+                // Since SymbolPackageUpdateResourceV3 will be requested this should be a v3 PackageSource
+                packageSource = new PackageSource(source)
+                {
+                    ProtocolVersion = 3
```
This doesn't feel safe to me. I started CI to see if any tests fail, but NuGet's tests aren't exhaustive, so even if all tests pass, I still won't be fully confident.

Specifically, if the customer is pushing to a V2 package source, this change will force NuGet into thinking it's a V3 package source, and then this line: https://github.com/NuGet/NuGet.Client/blob/c6e9542fcc9300361f03d5bf7fbab1732d4d19bf/src/NuGet.Core/NuGet.Commands/Utility/CommandRunnerUtility.cs#L119

instead of returning null, because v2 package sources don't support this resource type, NuGet will throw an exception.

Therefore, my gut feeling is that this will break pushing packages to any V2 package source. We'll see if any tests fail.

Zaneo replied:
It seemed to be returning null and not throwing an exception, that would have been preferable.

I'm ammenable to making `GetSymbolPackageUpdateResource` throw an exception if `packageSource = new PackageSource(source);` results in a V2 source.

Zaneo replied:
![image](https://github.com/NuGet/NuGet.Client/assets/295949/5d6d048e-d94e-4c7d-936a-bf8d2eaa552c)

Zaneo replied:
I cannot see anything in `PackageSource` constructor that sets `ProtocolVersion` based on the `source`.

Zaneo replied:
![image](https://github.com/NuGet/NuGet.Client/assets/295949/77411c2a-1877-4db6-a21a-8b9ac5c85fc6)

zivkan replied:
15 func tests are failing:

- CommandLine.FuncTest.Commands.PushCommandTest.PushCommand_Server_SkipDuplicate_IsSpecified_PushProceeds
- NuGet.CommandLine.FuncTest.Commands.PushCommandTest.PushCommand_Server_SkipDuplicate_NotSpecified_PushHalts
- NuGet.CommandLine.Test.NuGetPushCommandTest.PushCommand_LogsServerWarningsWhenPresent(firstServerWarning: "First of two server warning messages", secondServerWarning: "Second of two server warning messages")
- NuGet.CommandLine.Test.NuGetPushCommandTest.PushCommand_PushToServer_GetCredentialFromPlugin
- NuGet.CommandLine.Test.NuGetPushCommandTest.PushCommand_PushToServer_DoesNotDeadLockWhenSTDOutLarge
- NuGet.CommandLine.Test.NuGetPushCommandTest.PushCommand_PushToServerBasicAuth
- NuGet.CommandLine.Test.NuGetPushCommandTest.PushCommand_APIV2Package_Endpoint
- NuGet.CommandLine.Test.NuGetPushCommandTest.PushCommand_PushToServerIntegratedWindowsAuthenticationDisableBuffering
- NuGet.CommandLine.Test.NuGetPushCommandTest.PushCommand_PushToServer
- NuGet.CommandLine.Test.NuGetPushCommandTest.PushCommand_PushTimeoutErrorMessage
- NuGet.CommandLine.Test.NuGetPushCommandTest.PushCommand_PushToServerIntegratedWindowsAuthentication
- NuGet.CommandLine.Test.NuGetPushCommandTest.PushCommand_PushToServer_ApiKeyAsThirdArgument
- NuGet.CommandLine.Test.NuGetPushCommandTest.PushCommand_PushToServerFollowRedirection
- NuGet.CommandLine.Test.NuGetPushCommandTest.PushCommand_LogsServerWarningsWhenPresent(firstServerWarning: "Single server warning message", secondServerWarning: null)
- NuGet.CommandLine.FuncTest.Commands.PushCommandTest.PushCommand_AllowsTimeoutToBeSpecifiedHigherThan100Seconds

At least two of the tests failed with timeout errors, so might just be flakey, but a bunch of them failed with HTTP 404 errors, trying to get the service index, as I suspected.

NuGet.CommandLine is nuget.exe, and is ilmerged, so only builds on Windows. It's not clear from your screenshots if you're on Windows or another OS, so it might not be easy for you to run these tests. `NuGet.CommandLine.Tests` needs a nuget.exe built, and on Mac the tests run though mono-project.org. We don't run NuGet.CommandLine.Tests on Linux. I suspect NuGet.CommandLine.FuncTests probably also need a nuget.exe binary, but I'm less familiar with that test project.

Interestingly, none of the tests in `dotnet.integration.tests` failed, which tells me we don't have any dotnet CLI tests pushing to a mock V2 server in our tests. In fact, it looks like we don't have any push tests at all 😞 It should be possibly to copy one of the nuget.exe tests and port it to use the same test infrastructure as other dotnet cli tests. But this is non-trivial for a first time contributor.

Zaneo replied:
I'm building on Windows, and ran the UnitTests provide in the solution file but did not encounter any errors there. It might be a bit difficult as you say for me to take on porting all the tests though I think it would be good to have them.

I've changed the code to only set the `ProtocolVersion` to V3 if it ends in `index.json`, following the logic I found in: https://github.com/NuGet/NuGet.Client/blob/5ce332fb834feea1f6c974d17456c8c9d5504e09/src/NuGet.Clients/NuGet.VisualStudio.Common/Telemetry/TelemetryUtility.cs

Zaneo committed:
Conditially set the ProtocolVersion

This follows the same logic as in TelemetryUtility.IsHttpV3 if the source ends with index.json treat it as a V3 Protocol package source. That way there is a chance that resources can be fetched.

Zaneo committed:
Fix missing parenthesis

Zaneo commented:
I would like to add the following warning but I have two concerns:

1) If people have a workflow where they build symbol packages, then push to a V2 source, then a V3 source. The first push will generate a warning which they may not care about.

2) I think this change requiressy least two PRs? The strings in String.resx are compiled, so the builds would fail until I first merge a PR with the new string. Is the prefered order to merge the new string first in a PR. Or to do three PRs? (1) To merge the warning with hard coded text. (2) To merge the warning message to Strings.resx and (3) To change the hardcoded warning to use the String.resx entry.

https://github.com/NuGet/NuGet.Client/blob/55f4def65efe4a2b6dbf0aee8d699e7f29cc008c/src/NuGet.Core/NuGet.Commands/Strings.resx#L1038-L1041

(https://github.com/NuGet/NuGet.Client/blob/55f4def65efe4a2b6dbf0aee8d699e7f29cc008c/src/NuGet.Core/NuGet.Commands/Strings.resx#L1038-L1041)

Zaneo commented:
I would like to add the following warning but I have two concerns:

1) If people have a workflow where they build symbol packages, then push to a V2 source, then a V3 source. The first push will generate a warning which they may not care about.

2) I think this change requires least two PRs? The strings in String.resx are compiled, so the builds would fail until I first merge a PR with the new string. Is the prefered order to merge the new string first in a PR. Or to do three PRs? (1) To merge the warning with hard coded text. (2) To merge the warning message to Strings.resx and (3) To change the hardcoded warning to use the String.resx entry.

https://github.com/NuGet/NuGet.Client/blob/55f4def65efe4a2b6dbf0aee8d699e7f29cc008c/src/NuGet.Core/NuGet.Commands/Strings.resx#L1038-L1041

https://github.com/Zaneo/NuGet.Client/blob/55f4def65efe4a2b6dbf0aee8d699e7f29cc008c/src/NuGet.Core/NuGet.Commands/CommandRunners/PushRunner.cs#L54-L75

ghost commented:
This PR has been automatically marked as stale because it has no activity for **7 days**. It will be closed if no further activity occurs **within another 90 days** of this comment. If it is closed, you may reopen it anytime when you're ready again, as long as you don't delete the branch.
